### PR TITLE
[RAPTOR-5203] Add support for Julia Inference Models

### PR DESCRIPTION
**Pull request by timsetsfire**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary

Update DRUM to support inference models written in Julia.  Support is added by way of Pyjulia.  Julia models developed with the [MLJ framework](https://alan-turing-institute.github.io/MLJ.jl/dev/) and serialized with `.jlso` extensions are supported as drop in.  Additional modeling frameworks can be supported by way of custom.jl file provided necessary dependencies are installed.

### CODE REVIEW

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test functional_training_model_templates

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

File: custom_model_runner/setup.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
     "xgboost": extra_deps[SupportedFrameworks.XGBOOST],
     "R": ["rpy2;python_version>='3.6'"],
     "pypmml": extra_deps[SupportedFrameworks.PYPMML],
-    "trainingModels": ["datarobot==2.24.0"],
+    "trainingModels": ["datarobot==2.23.0b0"],
```
Comment (yakov-g) [inline custom_model_runner/setup.py:23]: Is this change intentional?

File: custom_model_runner/requirements.txt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 argcomplete==1.11.1
-datarobot==2.24.0
+datarobot==2.21.3
```
Comment (yakov-g) [inline custom_model_runner/requirements.txt:2]: Is this change intentional?

Comment (timsetsfire) [reply custom_model_runner/requirements.txt:2]: unintentional change.  will remedy

File: custom_model_runner/setup.py (hunk 1 above)
Comment (timsetsfire) [reply custom_model_runner/setup.py:23]: unintentional change - will remedy

File: custom_model_runner/requirements.txt (hunk 2 above)
Comment (yakov-g) [reply custom_model_runner/requirements.txt:2]: anyhow wait for complete reiveiw :)

**Comment by yakov-g** [comment]
jarvis test cmrunner_integration

**Review by yakov-g — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -197,6 +213,8 @@ Include any necessary hooks in a file called `custom.py` for Python models or `c
 Type checking methods can be used to verify types:
 - in Python`isinstance(data, str)` or `isinstance(data, bytes)`
 - in R `is.character(data)` or `is.raw(data)`
+- in Julia `data isa String` or `data is Base.CodeUnits`
```
[the remark this answers is not in the document]
Comment (timsetsfire) [reply README.md:216]: the command in julia is `isa`.

[the remark this answers is not in the document]
Comment (yakov-g) [reply README.md:216]: oh :)

File: model_templates/inference/julia/jl_custom/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+## Serving with Docker
+
+drum server --code-dir model_templates/inference/julia/jl_custom --target-type regression --input tests/testdata/boston_housing_inference.csv --verbose --logging-level info
+
+
+
+
+
+
```
[the remark this answers is not in the document]
Comment (timsetsfire) [reply model_templates/inference/julia/jl_custom/README.md:25]: on it 👍

File: model_templates/inference/julia/jl_unstructured/custom.jl [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        #, ret_kwargs
+    end
+    println("outgoing")
+    println(ret)
+    return ret
+end
+
+end
+
+# print("Model: ", model)
```
[the remark this answers is not in the document]
Comment (timsetsfire) [reply model_templates/inference/julia/jl_unstructured/custom.jl:51]: removed

File: public_dropin_environments/julia_mlj/dr_requirements.txt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+pyarrow==0.14.1
+datarobot-drum==1.4.16
```
[the remark this answers is not in the document]
Comment (timsetsfire) [reply public_dropin_environments/julia_mlj/dr_requirements.txt]: FIXED

File: public_dropin_environments/julia_mlj/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+numpy>=1.16.0,<1.19.0
+pandas==1.1.0
```
[the remark this answers is not in the document]
Comment (timsetsfire) [reply public_dropin_environments/julia_mlj/requirements.txt]: changed to 1.0.5 per comment.

File: public_dropin_environments/julia_mlj/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ENV PYTHON /usr/local/bin/python 
+RUN julia "$HOME/julia/sysim.jl" && \
+    chmod -R 777 $HOME/.julia/logs/manifest_usage.toml && \
+    rm julia-1.5.4-linux-x86_64.tar.gz
+
+ENV CODE_DIR=/opt/code 
+ENV ADDRESS=0.0.0.0:8080 WITH_ERROR_SERVER=1
+
+#Uncomment the following line to switch from Flask to uwsgi server
+ENV PRODUCTION=1 MAX_WORKERS=1 SHOW_STACKTRACE=1
```
[the remark this answers is not in the document]
Comment (timsetsfire) [reply public_dropin_environments/julia_mlj/Dockerfile:43]: FIXED

File: public_dropin_environments/julia_mlj/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,45 @@
+# This is the default base image for use with user models and workflows.
+# It contains a variety of common useful data-science packages and tools.
+FROM datarobot/python3-dropin-env-base
+
+# Install the list of core requirements, e.g. sklearn, numpy, pandas, flask.
+# **Don't modify this file!**
+COPY dr_requirements.txt dr_requirements.txt
```
[the remark this answers is not in the document]
Comment (timsetsfire) [reply public_dropin_environments/julia_mlj/Dockerfile:7]: This was a copy and paste of existing sklearn dropin.  Should it be changed in all?

File: README.md [hunk 10]
```diff
@@ -159,8 +174,8 @@ The model artifact must have a **jar** extension.
 Inference models support unstructured mode, where input and output are not verified and can be almost anything.
 This is your responsibility to verify correctness.
 
-### Custom hooks for Python and R models
-Include any necessary hooks in a file called `custom.py` for Python models or `custom.R` for R models alongside your model artifacts in your model folder:
+### Custom hooks for Python, R models, and Julia models
```
Comment (yakov-g) [inline README.md:177]: Probably there should be 3 `models` words or just 1.

**Comment by yakov-g** [comment]
We have an  option to provide language `--language`. When this option is used in `drum new` for the template generation, we exclude `java` from the available options as there is no template generation for java. Please  remove Julia as well (or you may want to add templates)
@staticmethod
    def _reg_arg_language(*parsers):
        for parser in parsers:
            langs = [e.value for e in RunLanguage]
            prog_name_lst = CMRunnerArgsRegistry._tokenize_parser_prog(parser)
            if prog_name_lst[1] == ArgumentsOptions.NEW:
                langs.remove(RunLanguage.JAVA.value)
                required_val = True
            else:
                required_val = False

            parser.add_argument(
                ArgumentsOptions.LANGUAGE,
                choices=langs,
                default=None,
                required=required_val,
                help="Language to use for the new model/env template to create",
            )

**Comment by yakov-g** [comment]
Go over over my comments, otherwise I approve.
Looks great!

**Review by yakov-g — APPROVED** [review]

File: custom_model_runner/setup.py (hunk 1 above)
Comment (timsetsfire) [reply custom_model_runner/setup.py]: FIXED

File: README.md (hunk 10 above)
Comment (timsetsfire) [reply README.md]: FIXED

**Review by rvorobii — APPROVED** [review]
